astra/src/main/java/com/slack/astra/graphApi/GraphBuilder.java

```diff
@@ -36,61 +40,224 @@ public class GraphBuilder {
   }
 
   /**
-   * Builds a dependency graph from a list of Zipkin spans.
+   * Filter for selecting nodes/edges in the graph based on metadata criteria.
    *
-   * <p>This method processes spans to create nodes (services) and edges (dependencies) representing
-   * the service communication graph. Each span becomes a node, and parent-child relationships
-   * between spans become edges in the graph. Logs warnings for any missing parent or child nodes.
+   * <p>The filter uses OR logic: a span matches if ANY of the filter criteria match. Each filter
+   * option is a field name (e.g., "operation", "service") mapped to a list of allowed values for
+   * that field.
+   *
+   * <p>Examples: {"operation": ["http.request"]} - matches spans with tag operation="http.request"
+   * {"operation": ["http.request", "grpc.request"]} - matches spans with either operation tag
+   * {"operation": ["http.request"], "kube.namespace": ["test-app-prod"]} - matches spans with
+   * operation="http.request" OR kube.namespace="test-app-prod" {} or null - empty filter matches
+   * all spans (no filtering)
+   *
+   * @param options Map of field names to lists of allowed values. If null or empty, all spans
+   *     match.
+   */
+  public record Filter(Map<String, List<String>> options) {
+    public boolean matches(ZipkinSpanResponse span) {
+      // Empty or null filter means match all spans
+      if (options == null || options.isEmpty()) {
+        return true;
+      }
+
+      // Returns true if ANY filter matches
+      return options.entrySet().stream()
+          .filter(entry -> entry.getValue() != null && !entry.getValue().isEmpty())
```

Comment:
you could move doing this filter into the constructor, instead of doing it each time